kernel/msg_q.c

```diff
@@ -178,7 +182,11 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
 
 	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, result);
 
-	k_spin_unlock(&msgq->lock, key);
+	if (resched) {
+		z_reschedule(&msgq->lock, key);
+	} else {
+		k_spin_unlock(&msgq->lock, key);
+	}
```

Comment:
FWIW I'm not a fan of this pattern, which is common but has been preserved only for compatibility.  Basically what this is saying is that "k_msgq_put() is a scheduling point if there was a thread waiting, **but not if there wasn't**", which is confusing to document and reason about.  (The difference in behavior being whether or not it will allow a third runnable higher priority thread to run or not).

Ideally an operation would be either a scheduling point or not, without this kind of logic.  Again we don't currently do that in many places because of history, and because changing scheduler behavior is often perceived as a compatibility bug.    But as we evolve it should be in that direction IMHO.

Reply:
Thanks for your explanation. It is very useful for me.